File: leetcode/24-swap-nodes-in-pairs.py
# https://leetcode.com/problems/swap-nodes-in-pairs/

from typing import Optional

# Definition for singly-linked list.
class ListNode:
	def __init__(self, val=0, next=None):
		self.val = val
		self.next = next

class Solution:
	def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
		"""
		Time Complexity : O(n)
		Space Complexity: O(1)
		"""
		if head == None or head.next == None:
			return head

		dummy = ListNode(0, head)

		first, second, third = dummy, head, head.next

		while third:
			# node swapping
			next = third.next
			third.next = second
			second.next = next
			first.next = third
			# pointer update
			first = first.next.next
			second = next
			third = next.next if next else None
		
		return dummy.next



	def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:

		def helper(node):
			if node == None or node.next == None:
				return node
			second = node.next
			node.next = helper(second.next)
			second.next = node
			return second

		return helper(head)



	def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
		"""
		Time Complexity : O(n)
		Space Complexity: O(n)
		"""
		if head == None or head.next == None:
			return head
		
		values = list()
		curr = head
		while curr:
			values.append(curr.val)
			curr = curr.next
			if len(values) % 2 == 0:
				values[-2], values[-1] = values[-1], values[-2]

		
		def listBuilder(index):
			if index >= len(values):
				return None
			swapedList = ListNode(values[index])
			swapedList.next = listBuilder(index+1)
			return swapedList

		return listBuilder(0)

File: leetcode/24-swap-nodes-in-pairs.js
// https://leetcode.com/problems/swap-nodes-in-pairs/


// Definition for singly-linked list.
function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function(head) {
  if (head === null || head.next === null) {
    return head;
  }

  let dummy = new ListNode();

  // window of the swapping operation
  let node0 = dummy;
  let node1 = head;
  let node2 = head.next;
  let node3 = head.next.next;

  // node0 -> node1 -> node2 -> node3 : we need a window of the 4 nodes envolved in the swapping operation
  // the two swapped nodes (node1 and node2) + one node before (node0) and one node after (node3)
  // we should swap node1 and node2 and make node0.next pointes to node2

  while (node2) {

    node1.next = node3;
    node2.next = node1;
    node0.next = node2;

    // shift our window of 4 nodes
    node0 = node1;
    node1 = node3;
    node2 = node3 && node3.next;
    node3 = node2 && node2.next;

  }

  return dummy.next;
};